source/_posts/2025-07-18-why-i-disallow-facade-usage.md

```diff
@@ -0,0 +1,333 @@
+</p>
+
+<p>
+    @TODO expose how this problem occurs especially when:
+     * an application grows in size, accommodating for multiple service instances
+     * tests: multiple services needed, one per test!
+</p>
+
+<h2>Problem 2: hidden dependencies</h2>
+
+<p>
+    A second problem occurs when running this code in isolation: assuming you have **only**
+    autoloading configured, what will it do?
+</p>
+
+~~~php
+<?php
+
+require_once __DIR__ . '/vendor/autoload.php'
+
+$service = new \My\Namespace\MyService();
+
+$result = $service->doSomeHeavyWork(); // this will crash: the cache was never configured!
+~~~
+
+<p>
+    This kind of crash is very similar to what you'd experience with the service location pattern: since
+    you are deferring all instantiation to the point at which a dependency is effectively invoked, you
+    may run in crashes due to a missed dependency.
+</p>
+
+<p>
+    In addition to the disadvantages of service-location, you also have the hidden dependency of the facade's
+    internal
+    <a href="https://github.com/laravel/framework/blob/2c682e4eb531eae6439579f25d29429ecc0a66ca/src/Illuminate/Support/Facades/Facade.php#L20-L25">
+        <code>static::$app</code>
+    </a>.
+</p>
+
+<p>
+    Your code cannot work until you've bootstrapped a Laravel application, which is a complex and heavy
+    (performance-wise) operation.
+</p>
+
+
+<h2>Problem 2: framework dependencies added</h2>
+
+<p>
+    Another issue with our <code>MyService</code> is that we have widely expanded the contact surface with the
+    framework: our code now depends on <code>illuminate/support</code>, and also on <code>illuminate/contracts</code>
+    in order for it to function at basic level.
+</p>
+
+<p>
+    This is both an issue of Laravel, which exposes a very bloated <code>illuminate/contracts</code> package, and
+    of our code, since framework upgrades can now affect our ability to perform upgrades.
+</p>
+
+<p>
+    Coming myself from the
+    <a href="https://en.wikipedia.org/wiki/Domain-driven_design"><abbr title="Domain-Driven Design">DDD</abbr></a>
+    community, isolating software dependencies is critical for the long-term maintenance of a system, and reducing
+    the amount of dependencies is always a good idea, as it is entropy that will easily spin out of control.
+</p>
+
+<p>
+    In this case, we wanted a cache, not the entire framework.
+</p>
+
+<h2>Problem 3: hidden complexity</h3>
+
+<p>
+    @TODO expose here how adding a facade introduces magic method calls, stack frames, static analysis complexity.
+</p>
+
+<p>
+    @TODO expose concept of "simple" != "easy". Systems can be complex and easy, or harder to use, but simple.
+</p>
+
+<h2>The simple solution</h2>
+
+<p>
+    Here's our service re-implemented to make things simple and explicit:
+</p>
+
+~~~php
+<?php
+
+final class MyService implements SomeContract {
+    public function __construct(private readonly SomeCache) {}
```

Comment:
```suggestion
    public function __construct(private readonly SomeCache $cache) {}
```

Reply:
As noted on Discord, I should also mention that this is "picked up by autowiring" by default